perf(autoware_kalman_filter): optimize `updateWithDelay` by replacing sparse matrix operations

k1832 opened the pull request:
## Description

### Summary
This PR optimizes the `TimeDelayKalmanFilter::updateWithDelay` function to improve computational performance and memory usage by removing unnecessary sparse matrix constructions. Benchmark results show a **~73% reduction** in execution time. Additionally, the unit test suite has been refactored to verify mathematical correctness against a ground-truth implementation.

### Changes
* **Removed Sparse Matrix Construction**: Removed the creation of the large, mostly zero extended measurement matrix (`C_ex`).
* **Block Arithmetic**: Replaced the global matrix multiplication `update(y, C_ex, R)` with direct Eigen block operations targeting specific state/covariance blocks.
* **Numerical Stability**: Implemented Cholesky decomposition (`LLT`) for the Kalman Gain calculation instead of standard inversion.
* **Memory Optimization**: Added `.noalias()` to matrix products to avoid unnecessary temporary allocations.
* **Test Suite Refactor**: Rewrote `test_time_delay_kalman_filter.cpp` from a monolithic test into a modular `gtest` fixture with comprehensive edge-case coverage.
* **Validation**: Added a check for dimension mismatches in the `C` matrix.

### Motivation
The previous implementation constructed a full-size matrix `C_ex` (dimension `dim_y` x `dim_x_ex`) which contained mostly zeros. Multiplying this against the full state covariance matrix `P_` resulted in significant unnecessary overhead. By isolating the specific blocks corresponding to the `delay_step`, we reduce the computational complexity significantly.

### Mathematical Equivalence
The logic remains mathematically equivalent to the standard Kalman update:
1.  **Innovation**: Instead of $y - C_{ex} x$, we compute $y - C x_{delayed}$.
2.  **Covariance ($S$)**: We utilize the sub-block of $P$ ($P_{dd}$) corresponding to the delay step, avoiding zero-multiplications.
3.  **Kalman Gain**: Solved via $S K^T = (P C^T)^T$ using LLT decomposition.

### Performance Benchmark
I measured the function latency using `bpftrace` during a logging simulation (based on the Autoware tutorial).

**Results:**
* **Average Latency:** Reduced from **~259,000 ns** to **~69,000 ns** (approx. 3.7x speedup).
* **Distribution:** The execution time distribution shifted significantly from the [128K, 256K] ns range to the [32K, 64K] ns range.

**Before Optimization:**
```text
@stats_ns: count 2924, average 245357, total 717424991
@stats_ns: count 2919, average 292559, total 853981688
@stats_ns: count 2922, average 240792, total 703594627
```

**After Optimization:**
```text
@stats_ns: count 2915, average 71015, total 207008979
@stats_ns: count 2919, average 67869, total 198110807
```

**Benchmark Command:**
```bash
# Path to the shared object
LIB="/home/$USER/autoware/install/autoware_kalman_filter/lib/libautoware_kalman_filter.so"
# Mangled function name
FUNC="_ZN8autoware13kalman_filter21TimeDelayKalmanFilter15updateWithDelayERKN5Eigen6MatrixIdLin1ELin1ELi0ELin1ELin1EEES6_S6_i"

sudo bpftrace -e "
uprobe:$LIB:$FUNC {
    @start[tid] = nsecs;
}
uretprobe:$LIB:$FUNC /@start[tid]/ {
    \$ns = nsecs - @start[tid];
    @stats_ns = stats(\$ns);
    @max_ns   = max(\$ns);
    @min_ns   = min(\$ns);
    @dist_ns  = hist(\$ns);
    delete(@start[tid]);
}"
```


## Related links
None.

## How was this PR tested?
I refactored the existing unit tests into a GTest fixture (`TimeDelayKalmanFilterTest`) to validate that the optimization breaks neither logic nor precision.

The new tests verify the implementation against **Ground Truth helper functions** (`ground_truth_predict`, `ground_truth_update`) which implement the standard, non-optimized matrix operations. The tests assert `isApprox` (epsilon `1e-5`) equality between the optimized TDKF class and the ground truth logic.

**Test Scenarios Covered:**
* **Standard Lifecycle:** Initialization, Prediction, and Update sequences.
* **Zero Delay Update:** Verifying behavior when `delay_step = 0` (equivalent to standard Kalman Filter).
* **Max Delay Update:** Verifying updates at the tail of the delay buffer (`kMaxDelayStep - 1`).
* **Error Handling:** Ensuring `false` is returned for invalid delay steps (negative or exceeding buffer size).
* **Multi-step Prediction:** Verifying state management when multiple predictions occur before a delayed update.

```bash
colcon test --packages-select autoware_kalman_filter --event-handlers console_direct+
```

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

k1832 committed:
Optimize `updateWithDelay` by replacing sparse matrix operations with block arithmetic

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

pre-commit-ci[bot] committed:
style(pre-commit): autofix

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

k1832 committed:
Add test

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

pre-commit-ci[bot] committed:
style(pre-commit): autofix

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

k1832 committed:
Update README.md for (Extended Kalman Filter)

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

pre-commit-ci[bot] committed:
style(pre-commit): autofix

k1832 committed:
Add more detailed comments

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

pre-commit-ci[bot] committed:
style(pre-commit): autofix

k1832 committed:
Handle negative delay step and refactor tests

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

k1832 committed:
Replace TDKF with full term

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

pre-commit-ci[bot] committed:
style(pre-commit): autofix

veqcc reviewed (approved):
Regarding the overall code structure and how the Pull Request (PR) is created, looks good to me.
I will leave the review of the specific algorithm and test content to the maintainers.

k1832 committed:
Remove LLM like comment

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

xmfcx commented:
- https://github.com/autowarefoundation/autoware_core/pull/776#issuecomment-3691300870 due to this PR, the required status checks are updated again and simplified.
- I'm very sorry that these changed again :bow: I've rebased this PR to pass the new checks.

k1832 committed:
Address PR review comments: add input validation, NaN guard, and docs fix

- Add dimension validation for y, R, and C matrices in updateWithDelay
- Add NaN/Inf check on Kalman gain to match base KalmanFilter::update
- Add [[unlikely]] branch hints on all error paths
- Clarify "diagonal block" comment terminology
- Fix README augmented state vector (x_{k-d} -> x_{k-d+1}) and clarify delay range
- Add test for C matrix dimension mismatch

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

pre-commit-ci[bot] committed:
style(pre-commit): autofix

k1832 committed:
fix: remove [[unlikely]] C++20 attributes for C++17 compatibility

The [[unlikely]] attribute is a C++20 feature that causes compilation
errors under clang-tidy which enforces strict C++17 compliance.

Signed-off-by: Keita Morisaki <kmta1236@gmail.com>

YamatoAndo committed:
Merge branch 'main' into optimize-ekf-update

k1832 committed:
Merge remote-tracking branch 'origin/main' into optimize-ekf-update
